### PR TITLE
Feature: thanos support

### PR DESCRIPTION
**Pull request by ekristen**
This allows a power user to enable thanos in the prometheus definition for the operator to deploy properly.

## Important

This does not implement thanos or solve how to route this data to thanos, this simply provides the ability to tell the prometheus operator to deploy the thanos sidecar so that it CAN be harvested with the appropriate setup.

## Changes

1. Had to rename http to `nginx-http` to prevent a conflict in the service discovery and port discovery because the prometheus operator does things a certain way.

## Overview

To enable thanos there's new values supported by the rancher-monitoring chart

```
thanos:
  enabled: true
```

You can configure the object storage yaml by putting [valid storage yaml](https://github.com/thanos-io/thanos/blob/master/docs/storage.md) for the property `objectConfig` 

**Note:** I named it `objectConfig` to not confuse it with `objectStorageConfig` which has a different set of properties required by the prometheus operator.

If thanos is enabled, the prometheus chart will now create a config map for the thanos config object storage.

```
thanos:
  enabled: true
  objectConfig:
    type: S3
    config:
      bucket: thanos
      region: us-east-1
      endpoint: minio.minio.svc.cluster.local:9000
      signature_version2: false
      insecure: true
      access_key: example
      secret_key: example
```

This adds the thanos image to the `prometheus` values so it can be set and changed by 

```
prometheus:
  image:
   thanos:
     repository: improbable/thanos
     tag: v0.6.0
```

## How To Use

Because the UI does not allow you to also modify by YAML you have to provide the thanos answers via the form.

```
prometheus.thanos.enabled=true
prometheus.thanos.objectConfig.type=S3
prometheus.thanos.objectConfig.config.bucket=thanos
prometheus.thanos.objectConfig.config.region=us-east-1
prometheus.thanos.objectConfig.config.endpoint=minio.minio.svc.cluster.local:9000
prometheus.thanos.objectConfig.config.signature_version2=false
prometheus.thanos.objectConfig.config.insecure=true
prometheus.thanos.objectConfig.config.access_key=example
prometheus.thanos.objectConfig.config.secret_key=example
```

### CODE REVIEW

**Comment by ekristen** [comment]
Here you go @drpebcak

File: charts/rancher-monitoring/v0.0.4/values.yaml [hunk 1]
```diff
@@ -366,6 +369,11 @@ prometheus:
     clusterDisplayName: ""
   cluster:
     alertManagerNamespace: ""
+  thanos:
```
Comment (drpebcak) [inline charts/rancher-monitoring/v0.0.4/values.yaml]: Are the duplicate repository and tag settings required for some reason?

Comment (ekristen) [reply charts/rancher-monitoring/v0.0.4/values.yaml]: That was a mistake. I will correct.

Comment (ekristen) [reply charts/rancher-monitoring/v0.0.4/values.yaml]: Fixed.

**Comment by drpebcak** [comment]
@cjellick lgtm, but probably needs this image mirrored into the rancher org, right?

**Comment by loganhz** [comment]
Hi @ekristen ,

How does your `thanos-query` query the thanos store API in each cluster? 
Do you need a node port or something in each cluster to expose the thanos store API even with the S3 storage enabled?

Just curious as we are also looking for a centralized monitoring solution. Maybe thanos is a good choice

**Comment by ekristen** [comment]
Thanos would be a great solution but that’s a whole separate stack or chart really. Im willing to share and talk in slack or email about it. 

This is about allowing the thanos sidecar run so if you have the thanos infrastructure setup you can grab it.

**Review by loganhz — DISMISSED** [review]

**Comment by ekristen** [comment]
Any chance this will make it in?

**Comment by ekristen** [comment]
This needs to be updated for v0.0.5 and found a few things that need to be tweaked to work better with the prometheus operator. This is still a bit value add, will update the PR soon.

**Comment by ekristen** [comment]
This has now been updated for 0.0.5. It would be nice to get this merged! It has zero impact on deploying rancher-monitoring if you do not pass in the `thanos.enabled` variable.

**Comment by cloudnautique** [comment]
@ekristen does this expose the Thanos endpoint externally? 

What additional configs are needed to leverage these enhancements with a user's own Thanos install?

**Comment by ekristen** [comment]
I need to update this PR a bit, I have an update that creates an access-thanos like the access-prometheus service that exposes the thanos sidecar on the cluster.

This does not expose it externally, this just tells the operator to deploy the thanos sidecar and providing the object storage information is provided properly, thanos will start uploading the TSDB parts to an S3 bucket. 

Actually deploying and using Thanos is a more complex endeavor. One that I've solved for my organization and am looking to see how I can share with others. It is complex depending on how your infrastructure is setup, and due to the top down type communication that thanos needs to query in real-time.

File: charts/rancher-monitoring/v0.0.5/values.yaml [hunk 2]
```diff
@@ -317,6 +317,9 @@ prometheus:
     proxy:
       repository: rancher/nginx
       tag: 1.17.4-alpine
+    thanos:
+      repository: improbable/thanos
+      tag: v0.6.0
```
Comment (rawmind0) [inline charts/rancher-monitoring/v0.0.5/values.yaml]: [`improbable/thanos` docker image repo](https://hub.docker.com/r/improbable/thanos) is marked as deprecated. Could you please update thanos image with new one??

```
-      repository: improbable/thanos
-      tag: v0.6.0
+      repository: thanosio/thanos
+      tag: v0.9.0
```

@cjellick , as [@drpebcak  mentioned](https://github.com/rancher/system-charts/pull/95#issuecomment-530576997)  Would be required to mirror thanos image in Rancher org??

**Review by rawmind0 — CHANGES_REQUESTED** [review]
Thanks for the PR @ekristen. 

As last `rancher-monitoring` chart version has been upgraded to 0.0.7, could you please rebase and apply PR to it??

**Comment by ekristen** [comment]
@rawmind0 do I need to also make these changes to 0.0.6 and 0.0.7 too or will rancher handle that?

**Comment by ekristen** [comment]
@rawmind0 do I need to undo 0.0.5? or just apply to both 0.0.6 and 0.0.7?

**Comment by ekristen** [comment]
Rebased sorta. I just branched `dev` again and made the changes to 0.0.7.

**Review by rawmind0 — DISMISSED** [review]

**Comment by rawmind0** [comment]
Thanks for the update @ekristen. 

LGTM

File: charts/rancher-monitoring/v0.0.5/values.yaml (hunk 2 above)
Comment (cloudnautique) [reply charts/rancher-monitoring/v0.0.5/values.yaml]: @rawmind0, yes, we should host the image on the Rancher org.

**Review by cloudnautique — DISMISSED** [review]

**Review by gitlawr — DISMISSED** [review]

File: charts/rancher-monitoring/v0.0.7/charts/prometheus/templates/secrets.yaml [hunk 3]
```diff
@@ -25,4 +25,20 @@ metadata:
...
+  name: {{ template "app.thanos.fullname" . }}
+  labels:
+    app: {{ template "app.name" . }}
+    chart: {{ template "app.version" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+stringData:
+  thanos.yaml: |-
+{{ toYaml .Values.thanos.objectConfig | indent 6 }}
+{{-end }}
```
Comment (spencergilbert) [inline charts/rancher-monitoring/v0.0.7/charts/prometheus/templates/secrets.yaml]: I think these needs to be:
```suggestion
{{- end }}
```

Comment (ekristen) [reply charts/rancher-monitoring/v0.0.7/charts/prometheus/templates/secrets.yaml]: Why?

Comment (ekristen) [reply charts/rancher-monitoring/v0.0.7/charts/prometheus/templates/secrets.yaml]: Nvm. I read the diff wrong. Fixing.

**Comment by spencergilbert** [comment]
It would be awesome to get this out in a 2.3.x release - would love to start using this as an enterprise customer :pray:

**Review by cloudnautique — APPROVED** [review]

**Comment by ekristen** [comment]
Thanks!

**Comment by spencergilbert** [comment]
Awesome! Thanks guys!

**Pull request merged**